Books: document p2p protocol (#216)
Cuprate/cuprate: books/protocol/src/p2p_network/common_types.md

| `type` | u8                                      | The address type |
| `addr` | An object whose fields depend on `type` | The address      |

#### IPv4

`type = 1`

| Fields   | Type | Description      |
| -------- | ---- | ---------------- |
| `m_ip`   | u32  | The IPv4 address |
| `m_port` | u16  | The port         |

#### IPv6

`type = 2`

| Fields   | Type                   | Description      |
| -------- | ---------------------- | ---------------- |
| `addr`   | [u8; 16] (epee string) | The IPv6 address |
| `m_port` | u16                    | The port         |

#### Tor

TODO:

#### I2p

TODO:

### Peer List Entry Base [^pl-entry-base] { #peer-list-entry-base }

| Fields                 | Type                                | Description                                                                                           |
| ---------------------- | ----------------------------------- | ----------------------------------------------------------------------------------------------------- |
| `adr`                  | [Network Address](#network-address) | The address of the peer                                                                               |

> | `id`                   | u64                                 | The random, self assigned, id of this node                                                            |
> | `last_seen`            | i64                                 | A field marking when this peer was last seen, although this is zeroed before sending over the network |
> | `pruning_seed`         | u32                                 | This peers pruning seed, `0` if the peer does no pruning                                              |
> | `rpc_port`             | u16                                 | This nodes RPC port, `0` if this peer has no public RPC port.                                         |

Reviewer: ```suggestion
| `id`                   | u64                                 | The random, self assigned ID of this node                                                            |
| `last_seen`            | i64                                 | A field marking when this peer was last seen, although this is zeroed before sending over the network |
| `pruning_seed`         | u32                                 | This peer's pruning seed, `0` if the peer does no pruning                                              |
| `rpc_port`             | u16                                 | This node's RPC port, `0` if this peer has no public RPC port.                                         |
```

| `rpc_credits_per_hash` | u32                                   | States how much it costs to use this node in credits per hashes, `0` being free           |

### Tx Blob Entry [^tb-entry] { #tx-blob-entry }

| Fields          | Type                   | Description                             |
| --------------- | ---------------------- | --------------------------------------- |
| `blob`          | bytes (epee string)    | The pruned tx blob                      |
| `prunable_hash` | [u8; 32] (epee string) | The hash of the prunable part of the tx |

### Block Complete Entry [^bc-entry] { #block-complete-entry }

| Fields         | Type                | Description                                               |
| -------------- | ------------------- | --------------------------------------------------------- |

> | `pruned`       | bool                | A bool for it the block is pruned                         |
> | `block`        | bytes (epee string) | The block blob                                            |
> | `block_weight` | u64                 | The blocks weight                                         |

Reviewer: ```suggestion
| `pruned`       | bool                | True if the block is pruned, false otherwise                         |
| `block`        | bytes (epee string) | The block blob                                            |
| `block_weight` | u64                 | The block's weight                                         |
```
